contrib/parseq-restli-client/src/main/java/com/linkedin/restli/client/ParSeqRestClient.java

```diff
@@ -194,13 +207,56 @@ static String generateTaskName(final Request<?> request) {
     }
   }
 
+  private <T> Task<Response<T>> withD2Timeout(final Task<Response<T>> task, ConfigValue<Long> timeout) {
+    if (timeout.getValue() != null && timeout.getValue() > 0) {
+      // transform TimeoutException to be compatible with current Task timeout exception message
+      String desc = timeout.getSource().map(src -> " src: " + src).orElse("");
+      String timeoutTaskName = "withTimeout " + Math.min(timeout.getValue(), Integer.MAX_VALUE)
+          + TimeUnitHelper.toString(TimeUnit.MILLISECONDS) + desc;
+      return task.transform(timeoutTaskName, (Try<Response<T>> tryGet) -> {
+        if (tryGet.isFailed() && tryGet.getError() instanceof TimeoutException) {
+          String timeoutExceptionMessage = "task: '" + task.getName() + "' " + timeoutTaskName;
+          return Failure.of(Exceptions.timeoutException(timeoutExceptionMessage));
+        } else {
+          return tryGet;
+        }
+      });
+    } else {
+      return task;
+    }
+  }
+
   private <T> Task<Response<T>> createTask(final String name, final Request<T> request,
       final RequestContext requestContext, RequestConfig config) {
     LOGGER.debug("createTask, name: '{}', config: {}", name, config);
+    return createTaskWithTimeout(name, request, requestContext, config);
+  }
+
+  // reconcile timeout specified in requestContext with the given timeout
+  private void reconcileRequestTimeout(final RequestContext requestContext, Long timeout) {
+    Number contextTimeout = (Number) requestContext.getLocalAttr(R2Constants.REQUEST_TIMEOUT);
+    int timeoutBound = (contextTimeout != null) ? contextTimeout.intValue() : Integer.MAX_VALUE;
```

Comment:
I think it could be simplified. Do we need to check with Integer.MAX_VALUE?
Isn't Math.min(timeout, contextTimeout) enough (except that it has to be properly cast to a common type).

Reply:
what happens if requestContext does not set R2Constants.REQUEST_TIMEOUT? In those case, we should  use the configured timeout.

Reply:
Simplified the code to make this easier to read.